csrc/fusion.h

```diff
@@ -147,7 +147,7 @@ class NVF_API Fusion : public IrContainer {
   void addInput(Val* input);
 
   //! Register output as an output of the fusion
-  void addOutput(Val* output);
+  void addOutput(Val* output, bool hide_output = false);
```

Comment:
How do you expect this to collaborate with Fusion::aliasOutputToInput? Fusion::aliasOutputToInput changes the hide_output bit in io_alias_.

Reply:
It does account for existing output though.

```
  io_alias_[output] = AliasInfo{
      .type = type,
      .aliased_io = input,
      .hide_output = !output->isFusionOutput()};

  if (!output->isFusionOutput()) {
    addOutput(output, true);
  }
```

I think I could add a couple more assert in `addOutput`.

aliasOutputToInput only needed to have the update in the output, I think the added logic PR ensures that, while still maintaining us having the *right* output entries in `outputs_`

Reply:
I do want to point out that `removeOutput` is not considering aliases yet. It's not too tricky to patch that. But we are also not using `removeOutput` in python API at the point. I can add that as a separate issue for follow up PR if we want to fix that.

Reply:
After reading https://github.com/NVIDIA/Fuser/issues/2564 (which I believe is the motivation for this PR), there's apparently a simpler fix that updates https://github.com/NVIDIA/Fuser/blob/baea1bb3b935f71cb1dead5737c6e6c572238393/csrc/python_frontend/fusion_record.h#L1397 to call addOutput before aliasOutputToInput when the user doesn't want to hide an output alias. Note that whether to hide_output is determined here: https://github.com/NVIDIA/Fuser/blob/baea1bb3b935f71cb1dead5737c6e6c572238393/csrc/fusion.cpp#L825. Wdyt?

Reply:
That'll work. But that makes integration code in thunder a bit awkward.

We parse a program linearly today. `copy_` is translated to `aliasOutputToInput`. So if we traverse a program topologically, we'll always write `aliasOutputToInput` before we call `addOutput`.

Granted that this can also be worked around by special casing for `copy_` in integration, but I prefer to not have integration (or nvfuser users) to live by those restrictions.